### ISSUE TITLE
Preserve default modes (e.g. vi-mode) in Lisp REPL

### ISSUE THREAD

**Issue by Ambrevar** [atlas-engineer/nyxt#793]
For users who have some modes like Emacs or VI by default, switching to lisp-repl will break their binding, in particular for switching buffers, which is embarassing.

This is quite a general problem: How do we decide which modes to keep by default?

I thought of the following: Create a "web-buffer" and a "repl-buffer" classes which would inherit from `buffer`, only specializing the `default-modes` slot.

Thus the user would add blocker-mode to the default modes in the `web-buffer` class and `vi-normal-mode` to the default modes in `buffer` (so that it applies everywhere).

@jmercouris Thoughts?

**Comment by jmercouris** [comment]
Maybe what we need to do is reorder the modes when they switch? That way they could at some point enter text, and at some point switch buffers?

**Comment by Ambrevar** [comment]
Sorry, I didn't understand what you mean with "when they switch".

**Comment by jmercouris** [comment]
When user switches between command/input mode, reorder modes so that the system works as expected.

**Comment by Ambrevar** [comment]
That's not the problem.  Right now when you fire up a REPL, neither
emacs-mode, vi-normal-mode nor vi-insert-mode is on, even if the user
enabled it by default.

Because of this, the user must first

    C-space vi-normal-mode

to re-enable VI bindings.  Then everything works as expected
(vi-normal-mode <-> vi-insert-mode switching works beautifully).

**Comment by jmercouris** [comment]
I see, if the default modes is only the problem, then it should suffice to change the constructor of the REPL mode to take this into account- it does have a reference to the buffer.

**Comment by Ambrevar** [comment]
Yup, but see my first post, I'm wondering if we could solve the more
general problem of mode classification.

Which modes make sense with a web view?  Without a web view?

We would need some sort of tree here (CLOS inheritance?).

By the way, you'll note that without a web view it's impossible to
search the REPL buffer...

We've got an opened can of worms we need to close :)

**Comment by Ambrevar** [comment]
Fixed by #927.
